language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/FromValueGenerator.scala

```diff
@@ -204,15 +292,20 @@ private[inner] object FromValueGenerator extends StrictLogging {
           extractor(param, valArg, CodeBlock.of("$L", valArg), args, packagePrefixes),
           orElseThrow(apiType, field),
         )
-
-      case TypePrim(PrimTypeContractId, _) =>
+      case TypePrim(PrimTypeContractId, ImmArraySeq(TypeVar(name))) =>
+        CodeBlock.of(
+          "fromValue$L.fromContractId($L.asContractId()$L.getValue())",
+          JavaEscaper.escapeString(name),
+          accessor,
+          orElseThrow(apiType, field),
+        )
+      case TypePrim(PrimTypeContractId, ImmArraySeq(_)) =>
         CodeBlock.of(
           "new $T($L.asContractId()$L.getValue())",
           javaType,
           accessor,
           orElseThrow(apiType, field),
         )
```

Comment:
This is what I meant by having different paths depending on whether you have a type variable. _You shouldn't try to change this_; it still makes sense to have different paths.

Reply:
Oh Do you mean I shouldn't tried to add the new `case TypePrim(PrimTypeContractId, ImmArraySeq(TypeVar(name)))`?  if I don't add this, fromContractId of the template valueDecoder will never be called? I am sure I have missed something.

Reply:
> I am sure I have missed something.

No, this is fine. You just ended up with different `javaType` values previously, which isn't good enough for this PR.